CIRC-265 - Instance-level hold for unavailable items

mtraneis opened the pull request:
## Purpose
This is a the continuation of implementing title-level requests, specifically providing the ability to place requests on unavailable copy by the instance ID.

## Approach

- Acquire the queues for each item
- Check the queues size. The copy associated with the smaller queue will be prioritized over others to have a request placed for it.
- If the queues' sizes are the same, compare the queues' last items' request expiration date. The item associated with the queue that has the shortest request expiration date is preferred.
(There have been discussions to use the checkout date of the copy. This approach could predict the request time more accurately, but not by much, while the implementation of it is not straightforward. We could revisit this tie-breaker criteria if necessary.)

In retrieving queues for each Item, I created a "lite" set of calls that would get back less data (and thus performing less calls to storage modules) for performance optimization because for this purpose it doesn't need the full set of attached data, such as Item. This optimization is not needed, but nice to have.

I tried to closely follow the call pattern that @marcjohnson-kint established in CIRC-245, namely, to attach RequestQueue to the Item and fetch it in ItemRepository. However, this was unsuccessful because it requires ItemRepository to have a reference of the RequestQueueRepository, and RequestQueueRepository makes use of the RequestRepository which needs to instantiate an ItemRepository instance, so we have a circular dependency. The attempt is seen in this commit: https://github.com/folio-org/mod-circulation/commit/2a4072273361b338711917b9c1d2b8bd7d59ea22
Instead I had to get the queues "on the side" in RequestByInstanceIdResource. 

I also borrowed the pattern used in CreateRequestService.createRequest() to pass around a wrapper data structure (i.e, RequestAndRelatedRecords) in the chained calls.

#### TODOS and Open Questions
<!-- OPTIONAL
- [ ] Use GitHub checklists. When solved, check the box and explain the answer.
-->

## Learning
<!-- OPTIONAL
  Help out not only your reviewer, but also your fellow developer!
  Sometimes there are key pieces of information that you used to come up
  with your solution. Don't let all that hard work go to waste! A
  pull request is a *perfect opportunity to share the learning that
  you did. Add links to blog posts, patterns, libraries or addons used
  to solve this problem.
-->

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

marcjohnson-kint commented:
> ## Purpose
> This is a the continuation of implementing title-level requests, specifically providing the ability to place requests on unavailable copy by the instance ID.
> 
> ## Approach
> * Acquire the queues for each item
> * Check the queues size. The copy associated with the smaller queue will be prioritized over others to have a request placed for it.
> * If the queues' sizes are the same, compare the queues' last items' request expiration date. The item associated with the queue that has the shortest request expiration date is preferred.
>   (There have been discussions to use the checkout date of the copy. This approach could predict the request time more accurately, but not by much, while the implementation of it is not straightforward. We could revisit this tie-breaker criteria if necessary.)
> 
> In retrieving queues for each Item, I created a "lite" set of calls that would get back less data (and thus performing less calls to storage modules) for performance optimization because for this purpose it doesn't need the full set of attached data, such as Item. This optimization is not needed, but nice to have.

I think there could be a clarity trade off as subsequent developers will need to decide between two different mechanisms for fetching a request queue. I wonder if there is a different way of achieving this flexibility.

> I tried to closely follow the call pattern that @marcjohnson-kint established in CIRC-245, namely, to attach RequestQueue to the Item and fetch it in ItemRepository. However, this was unsuccessful because it requires ItemRepository to have a reference of the RequestQueueRepository, and RequestQueueRepository makes use of the RequestRepository which needs to instantiate an ItemRepository instance, so we have a circular dependency. The attempt is seen in this commit: [2a40722](https://github.com/folio-org/mod-circulation/commit/2a4072273361b338711917b9c1d2b8bd7d59ea22)
> Instead I had to get the queues "on the side" in RequestByInstanceIdResource.
> 
> I also borrowed the pattern used in CreateRequestService.createRequest() to pass around a wrapper data structure (i.e, RequestAndRelatedRecords) in the chained calls.

I'm kinda glad that the pattern for fetching the queue in the item repository didn't work out, as the queue isn't something that is necessarily useful in all contexts where an item is used.

We could work around the circular dependency by refining the intermediary finders concept. Eventually I'd like defining a fetcher to be a single method implementation and usage to be a function reference (which would make smaller scale testing much easier) and repositories just become a convenient box for very common fetchers (and creation / update). 

The boolean parameters partially fulfil this, though they are a little unwieldy.

I'd really like to figure out a better approach than the wrapper data structure. That will likely need to wait until the structure of the code improves (or Java acquires tuples that can be easily destructured, which is other languages make this much nicer)

marcjohnson-kint commented on src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java:
```diff
@@ -206,16 +229,90 @@ private void createInstanceLevelRequests(RoutingContext routingContext) {
+      });
+  }
+
+  private static Map<Item, RequestQueue> sortRequestQueues(Map<Item,RequestQueue> unsortedItems){
```
```suggestion
  private static Map<Item, RequestQueue> sortRequestQueues(Map<Item, RequestQueue> unsortedItems) {
```

mtraneis commented on src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java:
```diff
@@ -206,16 +229,90 @@ private void createInstanceLevelRequests(RoutingContext routingContext) {
+      items.stream()
+        .filter(item -> !item.isAvailable())
+        .collect(Collectors.toList()));
+
```
Cool stuff! Thanks.

mtraneis commented on src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java:
```diff
@@ -206,16 +229,90 @@ private void createInstanceLevelRequests(RoutingContext routingContext) {
+        }
+
+        //Sort the map
+        Map<Item, RequestQueue> sortedMap = sortRequestQueues(itemQueueSizeMap);
```
Although it's more streamlined to return the keyset (a list of items), I would like to keep the method's signature this way as it more clearly defines the separation of responsibilities: an unsorted map in, sorted map out, in case some theoretical caller wants to use the method differently, and it makes unit testing it easier. I'll add more tests for this method.

marcjohnson-kint replied:
Are these maps used as maps, or are they just a handy way of keeping a pair of objects together?

marcjohnson-kint replied:
One aspect of using HashMaps is that they do not guarantee the [order of iteration](https://docs.oracle.com/javase/8/docs/api/java/util/HashMap.html)

marcjohnson-kint replied:
If we want to retain the map, another way to achieve this is:
```
    //Use linked hash map to guarantee sort order
    return unsortedItems
      .entrySet()
      .stream()
      .sorted(compareQueueLengths())
      .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue,
        (oldValue, newValue) -> oldValue, (LinkedHashMap::new)));
```

where `compareQueueLengths()` is the comparison extracted to a method.

```
  private static Comparator<Map.Entry<Item, RequestQueue>> compareQueueLengths() {
    return (q1, q2) -> {
      RequestQueue queue1 = q1.getValue();
      RequestQueue queue2 = q2.getValue();

      int result = queue1.size() - queue2.size();

      if (result == 0) {
        return queue1.getLowestPriorityFulfillableRequest()
          .getRequestExpirationDate()
          .compareTo(queue2.getLowestPriorityFulfillableRequest().getRequestExpirationDate());
      }

      return result;
    };
  }
```

mtraneis replied:
Yes, the use of hashmap here is to keep a pair of objects together with guaranteeing order.

mreno-EBSCO replied:
Do we even need a map? Doesn't each `Request` in each RequestQueue, assuming there are any, have an `itemId` that we could use? Maybe something we can do in the future to simplify this.

mtraneis replied:
I think we need to have a map, because as you said, the RequestQueue may not have a Request. I added a check for it last night and adding tests for this case now.

mreno-EBSCO replied:
My thought was for any item id in the sorted list of un/available items, if there is not a request queue that has that item id, then there are no requests for that item, so the length is 0. But I don't want to belabor this, since I have no idea if what I am saying is useful or would work. Since you are confident this already works, we can table this for a future discussion or just ignore it since we know this solution is not the ideal one. 😄

marcjohnson-kint commented on src/main/java/org/folio/circulation/domain/RequestQueueRepository.java:
```diff
@@ -48,6 +48,19 @@ public static RequestQueueRepository using(Clients clients) {
       .thenApply(r -> r.map(RequestQueue::new));
   }
 
+  public CompletableFuture<Result<RequestQueue>> getLiteRequestQueues(String itemId) {
```
does `lite` have a special meaning? It isn't a term I've encountered in a context like this before.

mreno-EBSCO replied:
I was a bit confused by this too. I thought it was a misspelling, but, after digging, I realized it was about less stuff being done. I blame Miller for the popularity of "lite" thanks to its Lite beer ads from the past 4 decades. Maybe rename this to `getRequestQueueWithoutItemLookup`. That's a mouthful. Initially, I was more confused about `findByLite` below. I thought "lite" was some sort of index to look up, something like `findByBarcode`.

mtraneis replied:
I like this, getRequestQueueWithoutItemLookup, @mreno-EBSCO

marcjohnson-kint commented on src/test/java/api/requests/InstanceRequestsAPICreationTests.java:
```diff
@@ -171,75 +180,292 @@ public void cannotSuccessfullyPlaceATitleLevelRequestWhenNoCopyIsAvailable()
+    List<IndividualResource> patrons2 = new ArrayList<>();
+    patrons2.add(usersFixture.steve());
+    patrons2.add(usersFixture.rebecca());
+    placeHoldRequest(item2, pickupServicePointId, patrons2, requestExpirationDate2);
```
Does this just create two hold requests?

It might be a stylistic difference, can you help me understand how this could be clearer than something like:
```
placePagedRequest(item1, pickupServicePointId, jessica, requestExpirationDate1);
placeHoldRequest(item2, pickupServicePointId, steve, requestExpirationDate2);
placeHoldRequest(item2, pickupServicePointId, rebecca, requestExpirationDate2);
```

mtraneis replied:
Yes, it creates two hold requests. This is another way of doing it, and it looks clearer.  I was thinking of loops over readability.

marcjohnson-kint commented on src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java:
```diff
@@ -206,16 +229,90 @@ private void createInstanceLevelRequests(RoutingContext routingContext) {
+
+      int result = queue1.size() - queue2.size();
+      if (result == 0) {
+        result = queue1.getLowestPriorityFulfillableRequest()
```
The issues references the `due date`, can you help me understand how that maps to expiration date?

mtraneis replied:
As alluded to earlier in the summary, neither approach works perfectly to predict the shortest wait time for the patron. I think requirement wise we have some wiggle room here. It's logical to compare the queue sizes, but then switching over to due date doesn't make sense to me. If a queue has 5 items but shorter due date on the current checkout item, but the other queue has 4 items and a much longer due date on its current checked out item, we'd pick the item with queue  size of 4.  Therefore I chose the last item's expiration date, but this isn't perfect. We could choose the minimum of all item's expiration date rather than the last item's. I had to make this choice because it wasn't ideal to ask Magda at 5am :) and it would require more code to get the loans up front.

marcjohnson-kint replied:
> neither approach works perfectly to predict the shortest wait time for the patron

Agreed, I'm only asking in order to understand what approach we have chosen to take.

> I think requirement wise we have some wiggle room here. It's logical to compare the queue sizes, but then switching over to due date doesn't make sense to me

Has this change in intended behaviour been discussed with Magda or Cate?

> I had to make this choice because it wasn't ideal to ask Magda at 5am :)

Ok, it sounds like it hasn't. Is she around now to ask?

If the proposed solution isn't viable, I'd prefer we just chose the first item, rather than invented an alternative approach. An improved approach can be added later.

mtraneis replied:
Yes, I discussed this with Magda afterward yesterday, and she was fine with it. She said 

> Whatever  we  pick the result will not be ideal and in many cases inaccurate because the patron can always return the item before their due date, the item with the shortest can go missing.  However from the data we found in survey conducted by RA SIG in most cases the queues are short if any. We also have other projects going in parallel that should alleviate the issue of  inaccuracy of calculating first available item.

marcjohnson-kint replied:
> We also have other projects going in parallel that should alleviate the issue of inaccuracy of calculating first available item.

Did she mention what those projects are?

marcjohnson-kint replied:
Ok, it's not my call, so I'll step away.

(Not relying on the contents of the queue could've given us options for how to do the processing to simplify the implementation)

mtraneis replied:
> (Not relying on the contents of the queue could've given us options for how to do the processing to simplify the implementation)

What other options do you have in mind?

marcjohnson-kint replied:
If we only cared about the length of the queue, we could only fetch that. Then it would be a number comparison only.

And any ties could be resolved with the first item or due date (as originally requested).

It would remove the need for the lighter way to fetch the queue, and probably some of the mapping and carrying of lists around. When we have faceting, it could likely be done with a single request.

Anyway, it isn't important any more given the interim decisions made.

mtraneis replied:
After more discussions we've decided to only care about the length of the queue because neither tie breaker approach discussed buys us much while introducing more code and adding more risks. Tie breaker is simply get the first item from the list, whichever it is.

marcjohnson-kint commented on src/main/java/org/folio/circulation/domain/RequestQueue.java:
```diff
@@ -28,6 +28,11 @@ Request getHighestPriorityFulfillableRequest() {
   }
 
+  public Request getLowestPriorityFulfillableRequest() {
+    return fulfillableRequests().get(fulfillableRequests().size() - 1);
```
What if there are no fulfillable requests in the queue?

mtraneis replied:
Yes, I added this check afterward and it will return null.

mtraneis replied:
This code isn't needed anymore. I removed it already.

marcjohnson-kint commented:
@mreno-EBSCO @mtraneis I'm going to leave it up to the two of you to decide whether to merge this (and whether to include anything from my experimental follow up branch)

mreno-EBSCO commented on src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java:
```diff
@@ -269,47 +261,54 @@ private void createInstanceLevelRequests(RoutingContext routingContext) {
 
+        if (failedQueuesItemList.size() == requestQueueFutures.size()) {
+          log.error("Failed to find request queues for all items of instanceId {}",
+                        ((List<Item>) unsortedUnavailableItems).get(0).getInstanceId());
```
Why does `InstanceRequestRelatedRecords.getUnsortedAvailableItems()` return a `Collection<Item>`? All other methods there and the member variables use `List<Item>`. I don't see any indication that we need to return a `Collection`. Changing to a `List<item>` would make the class consistent and we wouldn't have to cast here.

mtraneis replied:
Good catch. I was wondering about why the IDE automatically casted it for me as well...

mtraneis commented:
A couple of last-minutes changes made:
1. Removed the tie-breaker logic as it doesn't buy us much, if anything.
2. Added logic to handle items that have problems when retrieving request queues for them.  These items will be added to the end of the sorted list. They would only be used for Requests only when other items that were ranked before them didn't work out. 
- If it can't retrieve request queues for all unavailable items and there isn't any available item
